File: Easy/fil_drinks/fill_drinks.cpp
#include <iostream>
#include <algorithm>
using ll=long long;
using namespace std;
int main()
{
	ios::sync_with_stdio(false);
	cin.tie(nullptr);
	ll n;
	ll w1,w2,h1,h2;
	cin>>n>>w1>>w2>>h1>>h2;
	ll a1=w1*w1,a2=w2*w2;
	ll v1=a1*h1,v2=a2*h2;
	ll nowvol=0;
	ll nowheight=0;
	ll adh=0;
	for (ll i=0;i<n;i++)
	{
		ll vol; cin>>vol;
		nowvol=min(v1+v2,nowvol+vol);
		ll newheight;
		if (nowvol==v1+v2) 
			newheight=h1+h2;
		else if (nowvol>=v1)
			newheight=h1+(nowvol-v1)/a2;
		else
			newheight=nowvol/a1;
		adh=max(adh,newheight-nowheight);
		nowheight=newheight;
	}
	cout<<adh;
	return 0;
}

File: Easy/fil_drinks/fill_drinks.py
import sys
input=sys.stdin.readline
n=int(input())
w1,w2,h1,h2=map(int,input().split())
a1,a2=w1**2,w2**2
v1,v2=a1*h1,a2*h2
nowvol,nowheight=0,0
adh=0
for i in range(n):
	vol=int(input())
	nowvol=min(nowvol+vol,v1+v2)
	if (nowvol==v1+v2):
		newheight=h1+h2
	elif (nowvol>=v1):
		newheight=h1+(nowvol-v1)//a2
	else:
		newheight=nowvol//a1
	adh=max(adh,newheight-nowheight)
	nowheight=newheight
print(adh)